nixos/modules/services/web-apps/nextcloud.nix

```diff
@@ -190,13 +190,8 @@ in {
       description = lib.mdDoc "Which package to use for the Nextcloud instance.";
       relatedPackages = [ "nextcloud26" "nextcloud27" ];
     };
-    phpPackage = mkOption {
-      type = types.package;
-      relatedPackages = [ "php81" "php82" ];
-      defaultText = "pkgs.php";
-      description = lib.mdDoc ''
-        PHP package to use for Nextcloud.
-      '';
+    phpPackage = mkPackageOption pkgs "php" {
+      example = "php82";
```

Comment:
`relatedPackages` and the metadata associated with it is effectively thrown out of the manual now that such an option is recommended, correct?

Reply:
Right, I forgot about these. While going through the options, I was planning of taking them into consideration later.

Using `mkPackageOption` and specifying `relatedPackages` aren't mutually exclusive - it would probably be trivial to add an extra argument to `mkPackageOption`. 

However, I'm questioning their usecase. Consider these scenarios:

- There's two competing implementations of a package (e.g. heimdal and mit kerberos) -> One is default and the other is marked as an example. No need for `relatedPackages`
- There's different versions of a package (like php above) -> We specify the non-versioned alias as default (or at least `defaultText`) and provide a versioned one as example. The user will most likely understand that there are more versions to be found. `relatedPackages` is not needed.
- There's some sort of `extensions` or `plugins` option, that takes `listOf package`, and we'd like to show possible options -> Depending on the amount of related packages we have, maintaining such a list can become hard and kinda useless at a certain point. It's easy to forget to add an entry to a nixos module everytime you add a related package. It would be better to create package sets like `tmuxPlugins`. Smaller sets might outgrow the reasonable size for `relatedPackages` at some point.

Unless I've missed something, that only leaves the case of when there's 3 or more competing implementations for a module, which seems like a rare occurence.

As a side note, these don't appear on https://search.nixos.org.

If we are to keep them, I think it would be a good idea to go through all the options and make sure to specify it properly where it's relevant, and maybe add a small note in the contribution docs.

Reply:
>  it would probably be trivial to add an extra argument to mkPackageOption.

That's what bothers me about those shortcut functions a bit: if we add a little more, we'll just have another `mkOption`.

> We specify the non-versioned alias as default (or at least defaultText) and provide a versioned one as example. The user will most likely understand that there are more versions to be found. relatedPackages is not needed.

What if you have an interpreter (such as PHP in this example), but want to denote with this option which packages are compatible (e.g. the latest one or oldest one isn't).

> There's two competing implementations of a package (e.g. heimdal and mit kerberos) -> One is default and the other is marked as an example. No need for relatedPackages

For n=2 pkgs, sure.

> Depending on the amount of related packages we have, maintaining such a list can become hard and kinda useless at a certain point. It's easy to forget to add an entry to a nixos module everytime you add a related package. It would be better to create package sets like tmuxPlugins. Smaller sets might outgrow the reasonable size for relatedPackages at some point.

Well, it's not an enum where evaluation will fail with stuff not listed in there. Also, I think that's a case where it's better to describe that drvs from a given prefix are expected rather than listing every single one of them.

> As a side note, these don't appear on https://search.nixos.org.

I assume you mean the option search not the package search that appears by default?
But anyways, it does in the man-page and the HTML, if the nixos.org stuff doesn't, then it's a bug on their end :shrug:

Reply:
> That's what bothers me about those shortcut functions a bit: if we add a little more, we'll just have another mkOption.

That's true. But most of the common boilerplate will still be gone. Maybe it would be possible to reduce the maintaince cost of `mkPackageOption` (and similar) by passing a set of defaults to `mkOption` while also allowing freeform arguments, similar to how it's usually done with special builders that use `mkDerivation`?

> What if you have an interpreter (such as PHP in this example), but want to denote with this option which packages are compatible (e.g. the latest one or oldest one isn't).

I still think this option wouldn't be a good fit. By it's name `relatedPackages`, I wouldn't jump to the conclusion that there is a version restriction. It sounds like it's better suited for a `::: {.warning}` comment inside `extraDescription`.

> I assume you mean the option search not the package search that appears by default?
> But anyways, it does in the man-page and the HTML, if the nixos.org stuff doesn't, then it's a bug on their end 🤷

Yes, I meant the option search. I feel like the fact that it's missing and have not been reported as missing yet could be an indication that this option is not that useful as things are right now. I don't know whether that's due to underutilization or because it's usecase is genuinely limited. But as stated earlier, I suspect it's the latter.

Reply:
> But most of the common boilerplate will still be gone

I disagree here, I don't consider duplication/boilerplate necessarily bad.

`package` options don't have much that I'd expect to change soonish, but on the other hand you can configure your editor to implement a small snippet.

Here you need to remember the order of the arguments (and snippets won't help as much since you'd need to write special snippets for e.g. `extraDescription` etc.) where you have a single interface for `mkOption` itself. Also, I'm wondering how prone to misuse this is, a lot of people write e.g. `mkEnableOption "Whether to allow foo"` which results in `Whether to enable Whether to allow foo.` in the description.

But anyways, the decision was made already (even before this PR) and that's not a hill I'd like to die on.

> I wouldn't jump to the conclusion that there is a version restriction. It sounds like it's better suited for a ::: {.warning} comment inside extraDescription.

Fair point.

> that this option is not that useful as things are right now.

Correct. Upon reflecting here I also came to the conclusion that I'm probably a vast minority here, so I'll resolve this now and leave things as they are.

I'll probably resume writing `package = mkOption`, others will fix it up later and that's fine for me.

Anyways, thanks a lot for replying :)